### ISSUE TITLE
[F#] Implement new Concept Exercise: string-formatting

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#1185]
This issue describes how to implement the `string-formatting` concept exercise for the F# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

## Goal

The goal of this exercise is to teach the student the basics of the Concept of Integral Numbers in F#.

## Learning objectives

- Know how to use the `ToString()` method to convert any object to a `string`.
- Know how to use printf formatting on values of any type.
- Know what the benefit of printf formatting is over string concatenation.
- Know how to use default format strings to convert to standard output formats.
- Know how to use custom format strings to convert to custom output formats.
- Know how to combine strings using functions in the `String` module.
- Know of the `StringBuilder` type and when to use it.

## Out of scope

Nothing.

## Concepts

This Concepts Exercise's Concepts are:

- `string-formatting`: know how to use the `ToString()` method to convert any object to a `string`; know how to use printf formatting on values of any type; know what the benefit of printf formatting is over string concatenation; know how to use default format strings to convert to standard output formats; know how to use custom format strings to convert to custom output formats; know how to combine strings using functions in the `String` module; know of the `StringBuilder` type and when to use it.

## Prequisites

This Concept Exercise's prerequisites Concepts are:

- `strings`: strings will be formatted.
- `inheritance`: knowing that each class derives from `object` and thus has built-in methods.

Any data types used in this exercise (e.g. `datetimes`) should also be added as prerequisites.

### Hints

- [`String` module][string-module]: the `String` module.
- [Formatting text][formatting-text]: shows how to use the printf-family functions to format text.

The following links all use C# as their language, but are equally valid for F#:

- [Formatting types][formatting-types]: the available format types to use.
- [Standard numeric format strings][standard-numeric-format-strings]: lists the standard numeric format strings.
- [Custom numeric format strings][custom-numeric-format-strings]: describes how to define custom numeric format strings.
- [Standard date and time format strings][standard-date-and-time-format-strings]: lists the standard date and time format strings.
- [Custom date and time format strings][custom-date-and-time-format-strings]: describes how to define custom date and time format strings.
- [String builder][string-builder]: shows how to use the string builder.

### After

- [`String` module][string-module]: the `String` module.
- [Formatting text][formatting-text]: shows how to use the printf-family functions to format text.

The following links all use C# as their language, but are equally valid for F#:

- [Formatting types][formatting-types]: the available format types to use.
- [Standard numeric format strings][standard-numeric-format-strings]: lists the standard numeric format strings.
- [Custom numeric format strings][custom-numeric-format-strings]: describes how to define custom numeric format strings.
- [Standard date and time format strings][standard-date-and-time-format-strings]: lists the standard date and time format strings.
- [Custom date and time format strings][custom-date-and-time-format-strings]: describes how to define custom date and time format strings.
- [String builder][string-builder]: shows how to use the string builder.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific logic to be added to the [analyzer][analyzer].

## Implementing

To implement the concept exercise, the following files must be created:

<pre>
languages
└── fsharp
    └── exercises
        └── concept
            └── string-formatting
                ├── .docs
                |   ├── instructions.md
                |   ├── introduction.md
                |   ├── hints.md
                |   └── after.md
                ├── .meta
                |   ├── config.json
                |   |── design.md
                |   └── Example.fs
                ├── StringFormatting.fs
                ├── StringFormatting.fsproj
                └── StringFormattingTests.fs
</pre>

## Step 1: add .docs/introduction.md

This file contains an introduction to the concept. It should make the exercise's learning goals explicit and provide a short introduction with enough detail for the student to complete the exercise. The aim is to give the student just enough context to figure out the solution themselves, as research has shown that self-discovery is the most effective learning experience. Using the proper technical terms in the descriptions will be helpful if the student wants to search for more information. If the exercise introduces new syntax, an example of the syntax should always be included; students should not need to search the web for examples of syntax.

As an example, the introduction to a "strings" exercise might describe a string as just a "Sequence of Unicode characters" or a "series of bytes". Unless the student needs to understand more nuanced details in order to solve the exercise, this type of brief explanation (along with an example of its syntax) should be sufficient information for the student to solve the exercise.

## Step 2: add .docs/instructions.md

This file contains instructions for the exercise. It should explicitly explain what the student needs to do (define a method with the signature `X(...)` that takes an A and returns a Z), and provide at least one example usage of that function. If there are multiple tasks within the exercise, it should provide an example of each.

## Step 3: add .docs/hints.md

If the student gets stuck, we will allow them to click a button requesting a hint, which shows this file. This will not be a "recommended" path and we will (softly) discourage them using it unless they can't progress without it. As such, it's worth considering that the student reading it will be a little confused/overwhelmed and maybe frustrated.

The file should contain both general and task-specific "hints". These hints should be enough to unblock almost any student. They might link to the docs of the functions that need to be used.

The hints should not spell out the solution, but instead point to a resource describing the solution (e.g. linking to documentation for the function to use).

## Step 4: add .docs/after.md

Once the student completes the exercise they will be shown this file, which should provide them with a summary of what the exercise aimed to teach. This document can also link to any additional resources that might be interesting to the student in the context of the exercise.

These files are also all described in the [concept exercises document][docs-concept-exercises].

## Step 5: update languages/fsharp/config.json

An entry should be added to the track's `config.json` file for the new concept exercise:

```json
{
  ...
  "exercises": {
    "concept": [
      ...
      {
        "slug": "string-formatting",
        "uuid": "89a2bb15-21f6-4794-b2c4-e6cd9aaffd29",
        "concepts": ["string-formatting"],
        "prerequisites": ["strings", "inheritance"]
      }
    ]
  }
}
```

## Step 6: adding track-specific files

These files are specific to the F# track:

- `StringFormatting.fs`. the stub implementation file, which is the starting point for students to work on the exercise.
- `StringFormatting.fsproj`: the F# project file.
- `StringFormattingTests.fs`: the test suite.
- `.meta/Example.fs`: an example implementation that passes all the tests. The project file should _not_ include this file.

## Step 7: update the general concept document

Add the exercise to the concept's reference document's `## Implementations` section ([example][reference-example]). As there currently is no reference document, a new reference document should be added to `reference/concepts/string_formatting.md`.

## Step 8: updating list of implemented exercises

- Add the exercise to the [list of implemented exercises][implemented-exercises].

## Step 9: add .meta/design.md:

This file contains information on the exercise's design, which includes things like its goal, its teaching goals, what not to teach, and more ([example][meta-design]). This information can be extracted from this GitHub issue.

## Step 10: add .meta/config.json:

This file contains meta information on the exercise, which currently only includes the exercise's contributors ([example][meta-config.json]).

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/fsharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/reference/concepts/type_inference.md
[analyzer]: https://github.com/exercism/fsharp-analyzer
[representer]: https://github.com/exercism/fsharp-representer
[meta-design]: https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/numbers/.meta/design.md
[meta-config.json]: https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/numbers/.meta/config.json
[docs-concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[integral-number-types]: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/integral-numeric-types
[string-builder]: https://docs.microsoft.com/en-us/dotnet/standard/base-types/stringbuilder
[string-module]: https://msdn.microsoft.com/visualfsharpdocs/conceptual/core.string-module-%5bfsharp%5d
[formatting-text]: https://fsharpforfunandprofit.com/posts/printf/
[formatting-types]: https://docs.microsoft.com/en-us/dotnet/standard/base-types/formatting-types
[standard-numeric-format-strings]: https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings
[custom-numeric-format-strings]: https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-numeric-format-strings
[standard-date-and-time-format-strings]: https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings
[custom-date-and-time-format-strings]: https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings
